lib/improver/lapse_rate.py

```diff
@@ -0,0 +1,308 @@
+        y_coord = temperature_cube.coord(axis='y').name()
+
+        # Extract orography and land/sea mask data.
+        orography_data = next(orography_cube.slices([y_coord,
+                                                     x_coord])).data
+        land_sea_mask = next(land_sea_mask_cube.slices([y_coord,
+                                                        x_coord])).data
+        # Fill sea points with NaN values.
+        orography_data = np.where(land_sea_mask, orography_data, np.nan)
+
+        # Extract data array dimensions to define output arrays.
+        dataarray_shape = next(temperature_cube.slices(
+            ["realization", y_coord, x_coord])).data[0].shape
+        dataarray_size = dataarray_shape[0] * dataarray_shape[1]
+
+        # Array containing all of the subsections extracted from data array.
+        all_temp_subsections = np.zeros((dataarray_size, footprintarray_size))
+        all_orog_subsections = np.zeros((dataarray_size, footprintarray_size))
+
+        # Creates cube list to hold lapse rate data.
+        lapse_rate_cube_list = iris.cube.CubeList([])
+
+        slices_over_realization = temperature_cube.slices_over("realization")
+        for temp_slice in slices_over_realization:
+
+            # Create slice to store lapse rate values.
+            lapse_rate_slice = temp_slice
+
+            temperature_data = temp_slice.data
+
+            # Fill sea points with NaN values. Can't use Numpy mask since not
+            # recognised by "generic_filter" function.
+            temperature_data = np.where(land_sea_mask, temperature_data,
+                                        np.nan)
+
+            # cval is value given to points outside the array.
+            fnc = fnc_class(allbuffers=all_temp_subsections)
+            generic_filter(temperature_data, fnc.filter, size=footprint_len,
+                           mode='constant', cval=np.nan)
+
+            fnc = fnc_class(allbuffers=all_orog_subsections)
+            generic_filter(orography_data, fnc.filter, size=footprint_len,
+                           mode='constant', cval=np.nan)
+
+            # height_diff_mask is True for points where the height
+            # difference between the central point and its neighbours
+            # is > max_height_diff.
+            height_diff_mask = create_height_diff_mask(
+                all_orog_subsections, ind_central_point, max_height_diff)
+
+            # Mask points with extreme height differences as NaN.
+            all_orog_subsections = np.where(height_diff_mask, np.nan,
+                                            all_orog_subsections)
+            all_temp_subsections = np.where(height_diff_mask, np.nan,
+                                            all_temp_subsections)
+
+            # Loop through both arrays and find gradient of each subsection.
+            # The gradient indicates lapse rate - save into another array.
+            lapse_rate_array = np.zeros(dataarray_size)
```

Comment:
There are an awful lot of very large arrays created in this plugin. I accept that we are paying for vectorisation with a large memory overhead, but we will need to be aware of this in case if causes any problems.